### PR TITLE
Fix some bugs with zipfile serialization

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #32958 - Make zip serialization the default
 * **#32244 - Fix some bugs with zipfile serialization**

It includes the following changes:
* Split up tests so that we can test both serialization methods
    * Loading something within a buffer doesn't work anymore, so those tests are only on the old serialization method (it's possible but introduces a big slowdown since it requires a linear scan of the entire zipfile to find the magic number at the end)
* Call `readinto` on a buffer if possible instead of `read` + a copy
* Disable CRC-32 checks on read (there was some issue where miniz said the CRC was wrong but `zipinfo` and `unzip` said the zip file was fine)

Differential Revision: [D19418935](https://our.internmc.facebook.com/intern/diff/19418935/)

### CODE REVIEW

File: third_party/miniz-2.0.8/miniz.c [hunk 1]
```diff
@@ -4522,7 +4522,9 @@ void *mz_zip_reader_extract_file_to_heap(mz_zip_archive *pZip, const char *pFile
 mz_bool mz_zip_reader_extract_to_callback(mz_zip_archive *pZip, mz_uint file_index, mz_file_write_func pCallback, void *pOpaque, mz_uint flags)
 {
     int status = TINFL_STATUS_DONE;
+#ifndef MINIZ_DISABLE_ZIP_READER_CRC32_CHECKS
```
Comment (driazati) [inline third_party/miniz-2.0.8/miniz.c]: This is a fix that was upstreamed in [this PR](https://github.com/richgel999/miniz/pull/143)

File: torch/csrc/generic/serialization.cpp [hunk 2]
```diff
@@ -6,8 +6,11 @@
 #include <c10/cuda/CUDAGuard.h>
 #endif
 
+// save_save is necessary since the old eager format saved storages as
```
Comment (jamesr66a) [inline torch/csrc/generic/serialization.cpp]: `save_size`

File: third_party/miniz-2.0.8/miniz.c (hunk 1 above)
Comment (jamesr66a) [reply third_party/miniz-2.0.8/miniz.c]: Are we going to update miniz at some point?

**Review by jamesr66a — APPROVED** [review]
Not really all that familiar with this code but nothing stands out to me as awful

**Comment by Evpok** [comment]
The added warning is somewhat worrying, since what is suggested would break the possibility of pickling tensors as other Python objects (for instance for now I can pickle a tuple of tensors with a simple call to `pickle.dumps`). Was there a discussion for this choice?

**Comment by driazati** [comment]
Not really (feel free to open one and tag whoever), but the thinking was that PyTorch has its own serialization format that uses `pickle` but is not exactly `pickle` since it has some other stuff. `torch.save` will already let you serialize arbitrary Python objects so long as they are pickle-able, and provides the mechanism to make `Tensor`s pickle-able. Is there some use case you have that isn't served by `torch.save`?

File: third_party/miniz-2.0.8/miniz.c (hunk 1 above)
Comment (driazati) [reply third_party/miniz-2.0.8/miniz.c]: We could but I think the update would just be the same as this change (plus whatever has been added since)

**Comment by Evpok** [comment]
@driazati Oh, right, I didn't remember that `torch.save` worked for arbitrary Python objects too. Thanks!

**Comment by ppwwyyxx** [comment]
Yes, torch has its own serialization function that may work better, but that doesn't justify removing pickle support. There are a lot of reasons to still enable pickle, like
* it's what everyone in the python world use by default. third-party libraries that we can't change may still use pickle under the hood
* simpler APIs (no equivalent of `pkl.dumps` in torch)

**Comment by andyljones** [comment]
Echoing @ppwwyyxx; I frequently pass around torch storage as part of larger datastructures, and insisting on pickle means any generic code that serializes those data structures will now need to import torch, even if they have no use for torch's functionality other than `save`.

The pickle module is pretty powerful - [there's a whole VM down there in fact](https://docs.python.org/3/library/pickletools.html). What functionality does `torch.save` provide you can't hook into that?

**Comment by driazati** [comment]
`torch.save` uses pickle under the hood, it just has some special handling for saving tensor data. The pickle support here is just calling `torch.save` and storing the tensor as that series of bytes. It'd probably be best to define some real contract around this (i.e. what parts of `torch` are `pickle.dumps`-able, since only `Storage` has it right now).

Being able to treat PyTorch objects as a black box is a pretty strong use case, can one of you file a follow up issue with the points made here to remove that warning?

**Pull request closed without merging**